src/components/WebCLI/connection.ts

```diff
@@ -3,84 +3,128 @@ type Options = {
   messageCallback: (message: string) => void;
   onclose: () => void;
   onopen: () => void;
+  onerror: (error: Event | string) => void;
 };
 
 class Connection {
-  address: string;
-  _messageBuffer = "";
-  _timeout: number | null = null;
-  _messageCallback: (message: string) => void;
-  _wsOnOpen: () => void;
-  _wsOnClose: () => void;
-  _ws: WebSocket | null = null;
+  #address: string;
+  #messageBuffer = "";
+  #timeout: number | null = null;
+  #messageCallback: (message: string) => void;
+  #wsOnOpen: () => void;
+  #wsOnClose: () => void;
+  #wsOnError: (error: Event | string) => void;
+  #ws: WebSocket | null = null;
 
   constructor(options: Options) {
-    this._messageCallback = options.messageCallback;
-    this.address = options.address;
-    this._wsOnOpen = options.onopen;
-    this._wsOnClose = options.onclose;
+    this.#messageCallback = options.messageCallback;
+    this.#address = options.address;
+    this.#wsOnOpen = options.onopen;
+    this.#wsOnClose = options.onclose;
+    this.#wsOnError = options.onerror;
+  }
+
+  get address() {
+    return this.#address;
+  }
+
+  get websocket() {
+    return this.#ws;
   }
 
   connect() {
-    const ws = new WebSocket(this.address);
-    ws.onopen = this._wsOnOpen;
-    ws.onclose = this._wsOnClose;
-    ws.onmessage = this._handleMessage.bind(this);
-    this._ws = ws;
+    const ws = new WebSocket(this.#address);
+    ws.onopen = this.#wsOnOpen;
+    ws.onclose = this.#wsOnClose;
+    ws.onmessage = this.#handleMessage.bind(this);
+    ws.onerror = this.#wsOnError;
+    this.#ws = ws;
     return this;
   }
 
   send(message: string) {
-    this._ws?.send(message);
+    this.#ws?.send(message);
   }
 
   disconnect() {
-    this._ws?.close();
-    if (this._timeout) {
-      clearTimeout(this._timeout);
+    if (
+      this.#ws?.readyState === WebSocket.CONNECTING ||
+      this.#ws?.readyState === WebSocket.OPEN
+    ) {
+      this.#ws.close();
+    }
+    if (this.#timeout) {
+      clearTimeout(this.#timeout);
     }
   }
 
-  _handleMessage(e: MessageEvent) {
+  isOpen() {
+    return this.#ws?.readyState === WebSocket.OPEN;
+  }
```

Comment:
Also add the active check.

```suggestion
  isOpen() {
    return this.#ws?.readyState === WebSocket.OPEN;
  }
  
  isActive() {
    return this.isOpen() || this.#ws?.readyState === WebSocket.CONNECTING;
  }
```